Mathlib/FieldTheory/SeparableClosure.lean

```diff
@@ -0,0 +1,386 @@
+    fun h ↦ top_unique fun x _ ↦ (mem_separableClosure_iff F E).2 (((isSeparable_def _ _).1 h) x)⟩
+
+/-- If `K / E / F` is a field extension tower, then `separableClosure F K` is contained in
+`separableClosure E K`. -/
+theorem separableClosure.le_restrictScalars [Algebra E K] [IsScalarTower F E K] :
+    separableClosure F K ≤ (separableClosure E K).restrictScalars F := fun x hx ↦ by
+  simp only [mem_restrictScalars, mem_separableClosure_iff] at hx ⊢
+  exact hx.map_minpoly E
+
+/-- If `K / E / F` is a field extension tower, such that `E / F` is separable, then
+`separableClosure F K` is equal to `separableClosure E K`. -/
+theorem separableClosure.eq_restrictScalars_of_isSeparable [Algebra E K] [IsScalarTower F E K]
+    [IsSeparable F E] : separableClosure F K = (separableClosure E K).restrictScalars F := by
+  refine le_antisymm (separableClosure.le_restrictScalars F E K) fun x hx ↦ ?_
+  rw [mem_restrictScalars, mem_separableClosure_iff,
+    ← isSeparable_adjoin_simple_iff_separable] at hx
+  haveI : IsSeparable F (E⟮x⟯.restrictScalars F) := IsSeparable.trans F E E⟮x⟯
+  have h : x ∈ E⟮x⟯.restrictScalars F := mem_adjoin_simple_self E x
+  exact (mem_separableClosure_iff F _).2 <| separable_of_mem_isSeparable F _ h
+
+/-- If `K / E / F` is a field extension tower, then `E` adjoin `separableClosure F K` is contained
+in `separableClosure E K`. -/
+theorem separableClosure.adjoin_le [Algebra E K] [IsScalarTower F E K] :
+    adjoin E (separableClosure F K) ≤ separableClosure E K :=
+  adjoin_le_iff.2 <| le_restrictScalars F E K
+
+/-- A compositum of two separable extensions is separable. -/
+instance IntermediateField.isSeparable_sup (L1 L2 : IntermediateField F E)
+    [h1 : IsSeparable F L1] [h2 : IsSeparable F L2] :
+    IsSeparable F (L1 ⊔ L2 : IntermediateField F E) := by
+  rw [← le_separableClosure_iff] at h1 h2 ⊢
+  exact sup_le h1 h2
+
+/-- A compositum of separable extensions is separable. -/
+instance IntermediateField.isSeparable_iSup {ι : Type*} {t : ι → IntermediateField F E}
+    [h : ∀ i, IsSeparable F (t i)] : IsSeparable F (⨆ i, t i : IntermediateField F E) := by
+  simp_rw [← le_separableClosure_iff] at h ⊢
+  exact iSup_le h
+
+end separableClosure
+
+namespace Field
+
+/-- The (infinite) separable degree for a general field extension `E / F` is defined
+to be the degree of `(separableClosure F E) / F`. -/
+def sepDegree := Module.rank F (separableClosure F E)
+
+/-- The (infinite) inseparable degree for a general field extension `E / F` is defined
+to be the degree of `E / (separableClosure F E)`. -/
+def insepDegree := Module.rank (separableClosure F E) E
+
+/-- The (finite) inseparable degree for a general field extension `E / F` is defined
+to be the degree of `E / (separableClosure F E)` as a natural number. It is defined to be zero
+if such field extension is infinite. -/
+def finInsepDegree : ℕ := Cardinal.toNat (insepDegree F E)
```

Comment:
I wonder if `finrank (separableClosure F E) E` will be easier to work with? (it is definitionally the same)

Reply:
It's easier to `rw [finrank]` than `rw [← finrank]` so it's probably a good idea. If we keep the current def:
```suggestion
def finInsepDegree : ℕ := (insepDegree F E).toNat
```

Reply:
> I wonder if `finrank (separableClosure F E) E` will be easier to work with? (it is definitionally the same)

This change breaks some of the following proofs using `simp only`.

>  Suggested change
```lean
- def finInsepDegree : ℕ := Cardinal.toNat (insepDegree F E)
+ def finInsepDegree : ℕ := (insepDegree F E).toNat
```

Unfortunately this doesn't work:
```
invalid field notation, function 'Cardinal.toNat' does not have argument with type (Cardinal ...) that can be used, it must be explicit or implicit with a unique name
```

Reply:
I see, .toNat is a bundled ZeroHom. (Is it not a MonoidHom?) The CoeFun symbol no longer show up in the docs, which confused me.

Reply:
I'd suppose `finrank` has some more lemmas around it? You can add `finrank` to the `simp only`s to fix them if that is the only concern. Otherwise adding the equality as a rfl lemma might be useful as well.

Reply:
> I'd suppose `finrank` has some more lemmas around it? You can add `finrank` to the `simp only`s to fix them if that is the only concern. Otherwise adding the equality as a rfl lemma might be useful as well.

Let me try later.